tests/test_model_building.py

```diff
@@ -0,0 +1,164 @@
+    """
+    frame = (
+        pl.DataFrame(
+            {
+                "geography": ["USA", "PA", "USA", "PA", "USA", "PA", "USA", "PA"],
+                "date": [
+                    "2019-12-31",
+                    "2019-12-31",
+                    "2020-01-07",
+                    "2020-01-07",
+                    "2020-01-14",
+                    "2020-01-14",
+                    "2020-01-21",
+                    "2020-01-21",
+                ],
+                "estimate": [0.0, 0.0, 1.0, 0.1, 3.0, 0.3, 4.0, 0.4],
+                "season": ["2019/2020"] * 8,
+                "elapsed": [0, 0, 7, 7, 14, 14, 21, 21],
+                "interval": [None, None, 7, 7, 7, 7, 7, 7],
+            }
+        )
+        .with_columns(daily=(pl.col("estimate") / pl.col("interval")).fill_null(0))
+        .with_columns(previous=pl.col("daily").shift(1).over("geography"))
+        .with_columns(date=pl.col("date").str.strptime(pl.Date, "%Y-%m-%d"))
+    )
+
+    frame = iup.IncidentUptakeData(frame)
+
+    return frame
+
+
+def test_extract_starting_conditions(frame):
+    """
+    Use the last date for each grouping factor
+    """
+    output = iup.LinearIncidentUptakeModel.extract_starting_conditions(
+        frame, ("geography",)
+    )
+
+    assert output.sort("geography").equals(
+        pl.DataFrame(
+            {
+                "geography": ["USA", "PA"],
+                "last_date": [dt.date(2020, 1, 21), dt.date(2020, 1, 21)],
+                "last_daily": [4.0 / 7, 0.4 / 7],
+                "last_elapsed": [21, 21],
+                "last_cumulative": [8.0, 0.8],
+            }
+        ).sort("geography")
+    )
+
+
+def test_extract_standards(frame):
+    """
+    Make a dictionary with as many keys as provided variables,
+    and two values per sub-dictionary
+    """
+    output = iup.LinearIncidentUptakeModel.extract_standards(
+        frame, ("estimate", "elapsed")
+    )
+
+    assert len(output) == 2
+    assert len(output["estimate"]) == 2
+
+
+def test_fit(frame):
+    """
+    Model should fit a line to two points, giving a 'perfect' fit
+    """
+
+    output = iup.LinearIncidentUptakeModel()
+    output = output.fit(frame, ("geography",))
+
+    assert output.model.score(output.x, output.y) == 1.0
+
+
+def test_build_scaffold_handles_no_groups():
+    """
+    Set up a data frame with 5 rows and 7 columns if no grouping factors are given
+    """
+    start = pl.DataFrame(
+        {
+            "last_date": dt.date(2020, 1, 31),
+            "last_daily": 0.1,
+            "last_elapsed": 31,
+            "last_cumulative": 4.0,
+            "last_interval": 1,
+        }
+    )
+    start_date = dt.date(2020, 2, 1)
+    end_date = dt.date(2020, 2, 29)
+    interval = "7d"
+    group_cols = None
+
+    output = iup.LinearIncidentUptakeModel.build_scaffold(
+        start, start_date, end_date, interval, group_cols
+    )
+
+    assert output.shape[0] == 5
```

Comment:
you could write this as `output.shape == (5, 7)`

and this is another place that saying the exact data frame you expect would be good